### PR TITLE
fix(rust): fix several panics detected by cargo-fuzz

### PR DESCRIPTION
**Pull request by BaldDemian**
## Why?

Fix several new panics when feeding corner-case input found by cargo-fuzz

## What does this PR do?
- In `rust/README.md`, the right command to run all tests seems to be `cargo test --workspace`. Run `cargo test --features tests` will get: 
  Screenshot 2026-03-15 at 6 16 51 AM

- In `rust/fory-core/src/meta/type_meta.rs`, 
  - https://github.com/apache/fory/blob/5fc06f1db45337346db4ed380906c013f1e2f3f7/rust/fory-core/src/meta/type_meta.rs#L645
    will panic if `encoding_idx` exceeds the size of `encodings`.
  - https://github.com/apache/fory/blob/5fc06f1db45337346db4ed380906c013f1e2f3f7/rust/fory-core/src/meta/type_meta.rs#L836
    will cause OOM if `num_fields` is too large. I limit the max value of `num_fields` to `i16::MAX` since `field_id` is `i16`

- In `rust/fory-core/src/row/bit_util.rs`, use saturating_add/mul to prevent potential overflow panic. But would it be better to return error instead of saturating_add/mul ?🤔

- In `rust/fory-core/src/row/reader.rs`, direct access into slice using `[]` may cause out-of-bounds panic.
  
- In `rust/fory-core/src/serializer/collection.rs`, `rust/fory-core/src/serializer/map.rs` and `rust/fory-core/src/serializer/primitive_list.rs`, we should check the remaining bytes in the buffer **before** allocating `Vec`. This can also prevent OOM.

- In `rust/fory-core/src/serializer/skip.rs`, `generics.first().unwrap()` and `generics.get(1).unwrap()` will panic if the size of `generics` is not long enough.

## Related issues
N/A

## AI Contribution Checklist

N/A

## Does this PR introduce any user-facing change?

N/A

## Benchmark

This PR only adds additional check in case of corner-case input and thus won't has major influence on the performance.

### CODE REVIEW

File: rust/fory-core/src/row/bit_util.rs [hunk 1]
```diff
@@ -18,5 +18,5 @@
 const WORD_SIZE: usize = 8;
 
 pub fn calculate_bitmap_width_in_bytes(num_fields: usize) -> usize {
-    ((num_fields + 63) / 64) * WORD_SIZE
+    (num_fields.saturating_add(63) / 64).saturating_mul(WORD_SIZE)
```
Comment (chaokunyang) [inline rust/fory-core/src/row/bit_util.rs]: This change behaviour, I think you should let it crash. Such crash is basically programming error

File: rust/fory-core/src/row/reader.rs [hunk 2]
```diff
@@ -103,10 +121,17 @@ pub struct MapViewer<'r> {
...
     pub fn new(row: &'r [u8]) -> MapViewer<'r> {
-        let key_byte_size = LittleEndian::read_u64(&row[0..8]) as usize;
+        let Some(header) = row.get(0..8) else {
+            return MapViewer {
+                key_row: &[],
+                value_row: &[],
+            };
+        };
+        let key_byte_size = LittleEndian::read_u64(header) as usize;
+        let key_end = (8usize).saturating_add(key_byte_size).min(row.len());
```
Comment (chaokunyang) [inline rust/fory-core/src/row/reader.rs]: ditto here.

Comment (chaokunyang) [reply rust/fory-core/src/row/reader.rs]: For row format, it's not used for rpc. it's mostly used in data processing cases. So we don't need to handle malicous/invalid data.

**Comment by BaldDemian** [comment]
Thanks for your feedback, the mentioned changes are removed.

**Review by chaokunyang — APPROVED** [review]
LGTM
